accel-pppd/libnetlink/iputils.c

```diff
@@ -458,7 +463,78 @@ int __export ipaddr_del_peer(int ifindex, in_addr_t addr, in_addr_t peer)
 	return r;
 }
 
-int __export iproute_add(int ifindex, in_addr_t src, in_addr_t dst, in_addr_t gw, int proto, int mask, uint32_t prio)
+#ifdef HAVE_VRF
+static uint32_t ipvrf_get_table(const char *vrf_name)
+{
+	struct iplink_req {
+		struct nlmsghdr n;
+		struct ifinfomsg i;
+		char buf[4096];
+	} req;
+	struct rtnl_handle *rth = net->rtnl_get();
+	struct rtattr *tb[IFLA_MAX+1];
+	struct rtattr *li[IFLA_INFO_MAX+1];
+	struct rtattr *vrf_attr[IFLA_VRF_MAX + 1];
+	struct ifinfomsg *ifi;
+	int len;
+	uint32_t tb_id = RT_TABLE_MAIN;
+
+	log_ppp_info2("utils: getting route table for %s\n", vrf_name);
+
+	if (!vrf_name)
+		return tb_id;
+
+	memset(&req, 0, sizeof(req) - 4096);
+
+	req.n.nlmsg_len = NLMSG_LENGTH(sizeof(struct ifinfomsg));
+	req.n.nlmsg_flags = NLM_F_REQUEST;
+	req.n.nlmsg_type = RTM_GETLINK;
+	req.i.ifi_family = AF_UNSPEC;
+
+	addattr_l(&req.n, 4096, IFLA_IFNAME, vrf_name, strnlen(vrf_name, 512));
+
+	if (rtnl_talk(rth, &req.n, 0, 0, &req.n, NULL, NULL, 0) < 0) {
+		if (errno == ENODEV && !strncmp(vrf_name, "default", 8))
+			if (rtnl_rttable_a2n(&tb_id, "main"))
+				log_ppp_error(
+					"BUG: route table \"main\" not found.\n");
+		return tb_id;
+	}
+
+	ifi = NLMSG_DATA(&req.n);
+
+	len = req.n.nlmsg_len;
+
+	len -= NLMSG_LENGTH(sizeof(*ifi));
+	if (len < 0)
+		goto out;
+
+	parse_rtattr(tb, IFLA_MAX, IFLA_RTA(ifi), len);
+
+	if (!tb[IFLA_LINKINFO])
+		goto out;
+
+	parse_rtattr_nested(li, IFLA_INFO_MAX, tb[IFLA_LINKINFO]);
+
+	if (!li[IFLA_INFO_KIND] || !li[IFLA_INFO_DATA])
+		goto out;
+
+	if (strncmp(RTA_DATA(li[IFLA_INFO_KIND]), "vrf", 4))
```

Comment:
As I understand, this would be true if `vrf_name` is a name of a network link that exists in the system, but the link device is _not_ a VRF, right?

I feel that we shouldn't just silently ignore that and default to main. If something is trying to look up a non-existent VRF, shouldn't it be handled as an error in the RADIUS request?